### PR TITLE
Adding low-rank covariance

### PR DESCRIPTION
**Pull request by vivianwhite**
Adding Covariance module and low-rank covariance experiments

### CODE REVIEW

File: FactConv/pytorch_cifar.py [hunk 1]
```diff
@@ -32,6 +34,8 @@ def save_model(args, model):
                         help='filename for saved model')
 parser.add_argument('--seed', default=0, type=int, help='seed to use')
 parser.add_argument('--width', type=float, default=1, help='resnet width scale factor')
+parser.add_argument('--spatial_k', type=float, default=1, help='%spatial low-rank')
```
Comment (MuawizChaudhary) [inline FactConv/pytorch_cifar.py:36]: consider removing

File: FactConv/pytorch_cifar.py [hunk 2]
```diff
@@ -70,25 +74,29 @@ def save_model(args, model):
...
+wandb_dir = "/home/mila/v/vivian.white/scratch/v1-models/wandb"
 os.makedirs(wandb_dir, exist_ok=True)
 os.chdir(wandb_dir)
 
-run = wandb.init(project="FactConv", config=args,
-        group="pytorch_cifar", name=run_name, dir=wandb_dir)
+print("Num Learnable Params: ", sum(p.numel() for p in net.parameters() if
+    p.requires_grad))
+
+run = wandb.init(project="factconv", config=args, group="lowrankplusdiag", name=run_name, dir=wandb_dir)
```
Comment (MuawizChaudhary) [inline FactConv/pytorch_cifar.py]: might be worth logging the number of parameters

File: FactConv/pytorch_cifar.py [hunk 3]
```diff
@@ -70,25 +74,29 @@ def save_model(args, model):
...
 #wandb.watch(net, log='all', log_freq=1)
 
 
 criterion = nn.CrossEntropyLoss()
 optimizer = optim.SGD(net.parameters(), lr=args.lr,
                       momentum=0.9, weight_decay=5e-4)
-scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=200)
+scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer,\
+        T_max=args.num_epochs)
```
Comment (MuawizChaudhary) [inline FactConv/pytorch_cifar.py:106]: good job

File: RSN_experiments/LearnableCov.py [hunk 4]
```diff
@@ -1,201 +0,0 @@
-import torch
-from torch import Tensor
```
Comment (MuawizChaudhary) [inline RSN_experiments/LearnableCov.py]: why this this script removed here?

**Review by MuawizChaudhary — COMMENTED** [review]
just a high level overview, i'll do additional passes.

File: FactConv/conv_modules.py [hunk 5]
```diff
@@ -17,6 +18,148 @@ def _contract(tensor, matrix, axis):
 
 
 class FactConv2d(nn.Conv2d):
+    def __init__(
```
Comment (MuawizChaudhary) [inline FactConv/conv_modules.py:23]: This entire script, your re-using alot of code. I wonder if you could subclass FactConv extensions using this factconv module.

Comment (MuawizChaudhary) [reply FactConv/conv_modules.py:23]: as you mentioned, eventually we should refactor the other fact convs to use the covariance module.

File: FactConv/cov.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
+import torch
+from torch import Tensor
+import torch.nn as nn
+from torch.nn.parameter import Parameter
+from torch.nn.common_types import _size_2_t
+from typing import Optional, List, Tuple, Union
+import math
+
+class Covariance(nn.Module):
```
Comment (MuawizChaudhary) [inline FactConv/cov.py:9]: theres some repeated code here too. I wonder if we could condense this down.

File: FactConv/models/function_utils.py [hunk 7]
```diff
@@ -144,19 +191,31 @@ def _replace_layers_fact_with_conv(module):
...
-            for name, param in module.named_parameters():
-                if covariance == "channel":
-                    if "tri1_vec" in name:
-                        param.requires_grad = False
-                if covariance == "spatial":
-                    if "tri2_vec" in name:
-                        param.requires_grad = False
+        if isinstance(module, FactConv2d) or isinstance(module, DiagFactConv2d)\
+        or isinstance(module, DiagChanFactConv2d) or isinstance(module,LowRankFactConv2d)\
+        or isinstance(module, LowRankPlusDiagFactConv2d):
```
Comment (MuawizChaudhary) [inline FactConv/models/function_utils.py]: if we had all these factconv variations subclassing fact cov you could just check if it was a factcov instance.

File: FactConv/models/function_utils.py [hunk 8]
```diff
@@ -144,19 +191,31 @@ def _replace_layers_fact_with_conv(module):
...
+            for name, mod in module.named_modules():
+                if isinstance(mod, Covariance):
```
Comment (MuawizChaudhary) [inline FactConv/models/function_utils.py]: actually i think is better to remove the fact conv check and just check if its a covariance and disable gradients that way.

File: RSN_experiments/LearnableCov.py (hunk 4 above)
Comment (vivianwhite) [reply RSN_experiments/LearnableCov.py]: good question. appears to be a mistake i made during rebasing

**Comment by kamdh** [comment]
Are we ready to merge?
